File: typo-tester-kotlin/src/main/kotlin/adventureworks/userdefined/FirstName.kt
package adventureworks.userdefined

import typo.runtime.PgText
import typo.runtime.PgType
import typo.runtime.PgTypes
import typo.runtime.internal.arrayMap

data class FirstName(val value: String) {
    companion object {
        val pgText: PgText<FirstName> = PgText.instance { v, sb -> PgText.textString.unsafeEncode(v.value, sb) }
        val pgType: PgType<FirstName> = PgTypes.text.bimap(::FirstName) { it.value }
        val pgTypeArray: PgType<Array<FirstName>> = PgTypes.textArray.bimap(
            { arr -> arrayMap.map(arr, ::FirstName, FirstName::class.java) },
            { arr -> arrayMap.map(arr, { it.value }, String::class.java) }
        )
    }
}

File: typo-tester-typo-scala-old/src/scala/adventureworks/userdefined/FirstName.scala
package adventureworks.userdefined

import typo.dsl.Bijection
import typo.runtime.{PgText, PgTypes}

case class FirstName(value: String) extends AnyVal

object FirstName {
  given bijection: Bijection[FirstName, String] = Bijection.apply[FirstName, String](_.value)(FirstName.apply)
  given pgType: typo.runtime.PgType[FirstName] = PgTypes.text.bimap(FirstName.apply, _.value)
  given pgText: PgText[FirstName] = new PgText[FirstName] {
    override def unsafeEncode(v: FirstName, sb: java.lang.StringBuilder): Unit = PgText.textString.unsafeEncode(v.value, sb)
    override def unsafeArrayEncode(v: FirstName, sb: java.lang.StringBuilder): Unit = PgText.textString.unsafeArrayEncode(v.value, sb)
  }
}
